Netkan/Services/CachingHttpService.cs

```diff
@@ -0,0 +1,141 @@
+
+            _curl = new CurlEasy { UserAgent = CKAN.Net.UserAgentString };
+
+            var caBundle = ResolveCurlCaBundle();
+            if (caBundle != null)
+            {
+                _curl.CaInfo = caBundle;
+            }
+        }
+
+        public string DownloadPackage(Uri url, string identifier)
+        {
+            EnsureNotDisposed();
+
+            return _cache.GetCachedFilename(url) ?? 
+                _cache.Store(url, CKAN.Net.Download(url), string.Format("netkan-{0}", identifier), move: true);
+        }
+
+        public string DownloadText(Uri url)
+        {
+            EnsureNotDisposed();
+
+            Log.DebugFormat("About to download {0}", url);
+
+            var content = string.Empty;
+
+            _curl.Url = url.ToString();
+            _curl.WriteData = null;
+            _curl.WriteFunction = delegate(byte[] buf, int size, int nmemb, object extraData)
+            {
+                content += Encoding.UTF8.GetString(buf);
+                return size * nmemb;
+            };
+
+            var result = _curl.Perform();
+
+            if (result != CurlCode.Ok)
+            {
+                throw new Exception("Curl download failed with error " + result);
+            }
+
+            Log.DebugFormat("Download from {0}:\n\n{1}", url, content);
+
+            return content;
+        }
+
+        /// <summary>
+        /// Resolves the location of the cURL CA bundle file to use.
+        /// </summary>
+        /// <returns>The absolute file path to the bundle file or null if none is found.</returns>
+        private static string ResolveCurlCaBundle()
+        {
+            const string caBundleFileName = "curl-ca-bundle.crt";
+            const string ckanSubDirectoryName = "CKAN";
+
+            var bundle = new[]
+            {
+                // Working Directory
+                Environment.CurrentDirectory,
+
+                // Executable Directory
+                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
+
+                // %LOCALAPPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), ckanSubDirectoryName),
+
+                // %APPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), ckanSubDirectoryName),
+
+                // %PROGRAMDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), ckanSubDirectoryName),
+            }
+            .Select(i => Path.Combine(i, caBundleFileName))
+            .FirstOrDefault(File.Exists);
+
+            Log.InfoFormat("Using curl-ca bundle: {0}",bundle ?? "(none)");
+
+            return bundle;
+        }
+    }
+
+    internal sealed partial class CachingHttpService
+    {
+        private bool _isDisposed;
+
+        ~CachingHttpService()
+        {
+            Dispose(disposing: false);
```

Comment:
I'm not as well versed in C# destructors as I'd like to be. To check my understanding here:

- If  `Dispose` is called directly (eg, by leaving a `using` scope), then we run our curl clean-up, and set `_isDisposed` to true.
- If our `CachingHttpService` is destroyed, we don't touch curl and just set `_isDisposed` to true.
- We can call `EnsureNotDisposed()` to make sure our object hasn't been disposed (but presumably hasn't been destroyed, because if it's destroyed we couldn't have a reference to it).

So mostly I'm checking under what circumstances we'd see the destructor called *without* `Dispose` also being invoked just before it. I feel like I'm missing something with the logic flow here. (I don't know 100% how the .NET garbage collector works, but I presume it won't go destroying something you have an active reference to.)